indexing-service/src/main/java/io/druid/indexing/overlord/RemoteTaskRunner.java

```diff
@@ -1091,32 +1091,42 @@ private void removeWorker(final Worker worker)
   private void scheduleBlackListedNodesCleanUp()
   {
     ScheduledExecutors.scheduleAtFixedRate(
-            cleanupExec,
-            Period.ZERO.toStandardDuration(),
-            config.getWorkerBlackListCleanupPeriod().toStandardDuration(),
-            new Runnable()
-            {
-              @Override
-              public void run()
-              {
-                long currentTimeStamp = System.currentTimeMillis();
-                for (ZkWorker zkWorker : blackListedWorkers) {
-                  cleanBlackListedNode(zkWorker, currentTimeStamp);
-                }
-              }
+        cleanupExec,
+        Period.ZERO.toStandardDuration(),
+        config.getWorkerBlackListCleanupPeriod().toStandardDuration(),
+        () -> {
+          long currentTimeStamp = System.currentTimeMillis();
+          boolean shouldRunPendingTasks = false;
+
+          // must be synchronized while iterating:
+          // https://docs.oracle.com/javase/8/docs/api/java/util/Collections.html#synchronizedSet-java.util.Set-
+          synchronized (blackListedWorkers) {
+            for (ZkWorker zkWorker : blackListedWorkers) {
+              shouldRunPendingTasks |= cleanBlackListedNode(zkWorker, currentTimeStamp);
             }
+          }
+
+          if (shouldRunPendingTasks) {
+            runPendingTasks();
+          }
+        }
     );
   }
 
-  public void cleanBlackListedNode(ZkWorker zkWorker, long currentTimeStamp)
+  public boolean cleanBlackListedNode(ZkWorker zkWorker, long currentTimeStamp)
   {
-    // Clean blacklisted workers if blacklisted time has elapsed
-    if(currentTimeStamp - zkWorker.getLastCompletedTaskTime().getMillis() >
-            config.getWorkerBlackListBackoffTime().toStandardDuration().getMillis()){
-      // White listing node
-      log.info("Whitelisting worker [%s]. ", zkWorker);
-      blackListedWorkers.remove(zkWorker);
+    long remainingMillis = zkWorker.getLastCompletedTaskTime().getMillis()
```

Comment:
@gianm I submitted a revision that changed the behavior for these 3 questions as follows:
- Blacklist period starts from when a task failure triggers the blacklisting and not after the last completed task
- A successful task removes a worker from the blacklist
- A worker removed from the blacklist has its consecutive failure counter reset